packages/lib-utils/src/k8s/hooks/useK8sWatchResources.ts

```diff
@@ -0,0 +1,184 @@
+import { Map as ImmutableMap } from 'immutable';
+import * as React from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import * as k8sActions from '../../app/redux/actions/k8s';
+import { getActiveCluster } from '../../app/redux/reducers/core';
+import type { K8sModelCommon } from '../../types/k8s';
+import type { K8sState, SDKStoreState } from '../../types/redux';
+import {
+  transformGroupVersionKindToReference,
+  getReferenceForModel,
+  getGroupVersionKindForReference,
+} from '../k8s-utils';
+import type { GetWatchData } from './k8s-watch-types';
+import { getWatchData, getReduxData, NoModelError } from './k8s-watcher';
+import { useDeepCompareMemoize } from './useDeepCompareMemoize';
+import { useModelsLoaded } from './useModelsLoaded';
+import { usePrevious } from './usePrevious';
+import type {
+  ResourcesObject,
+  WatchK8sResources,
+  WatchK8sResultsObject,
+  WatchK8sResults,
+} from './watch-resource-types';
+
+/**
+ * Hook that retrieves the k8s resources along with their respective status for loaded and error.
+ * @param initResources resources need to be watched as key-value pair, wherein key will be unique to resource and value will be options needed to watch for the respective resource.
+ * @return A map where keys are as provided in initResouces and value has three properties data, loaded and error.
+ * @example
+ * ```ts
+ * const Component: React.FC = () => {
+ *   const watchResources = {
+        'deployment': {...},
+        'pod': {...}
+        ...
+      }
+ *   const {deployment, pod} = useK8sWatchResources(watchResources)
+ *   return ...
+ * }
+ * ```
+ */
+export const useK8sWatchResources = <R extends ResourcesObject>(
+  initResources: WatchK8sResources<R>,
+): WatchK8sResults<R> => {
+  const cluster = useSelector((state: SDKStoreState) => getActiveCluster(state));
+  const resources = useDeepCompareMemoize(initResources, true);
+  const modelsLoaded = useModelsLoaded();
+
+  const allK8sModels = useSelector<SDKStoreState, ImmutableMap<string, K8sModelCommon>>(
+    (state: SDKStoreState) => state.k8s?.getIn(['RESOURCES', 'models']),
+  );
+
+  const prevK8sModels = usePrevious(allK8sModels);
+  const prevResources = usePrevious(resources);
+
+  // Ref to immutable map mapping model's kind (string) to the model object (ImmutableMap)
+  const k8sModelsRef = React.useRef<ImmutableMap<string, K8sModelCommon>>(ImmutableMap());
+
+  if (
+    prevResources !== resources ||
```

Comment:
This conditional is _huge_. I know we migrated it from console and it is working, but I'd like to move some of the `|| `and `&&` and `.some` logic to some variables for readability.

Reply:
@vidyanambiar let's follow up on this. Could you create a ticket for cleanup?

Reply:
Here's the ticket, Bryan. I created it as a story but please let me know if it's better to have it as a task.
https://issues.redhat.com/browse/HAC-1162